kernel/src/log_compaction/writer.rs

```diff
@@ -0,0 +1,177 @@
+}
+
+impl RetentionCalculator for LogCompactionWriter {
+    fn table_properties(&self) -> &TableProperties {
+        self.snapshot.table_properties()
+    }
+}
+
+impl LogCompactionWriter {
+    pub fn try_new(
+        snapshot: Arc<Snapshot>,
+        start_version: Version,
+        end_version: Version,
+    ) -> DeltaResult<Self> {
+        if start_version > end_version {
+            return Err(Error::generic(format!(
+                "Invalid version range: start_version {} > end_version {}",
+                start_version, end_version
+            )));
+        }
+
+        // Compute the compaction path once during construction
+        let compaction_path =
+            ParsedLogPath::new_log_compaction(snapshot.table_root(), start_version, end_version)?;
+
+        Ok(Self {
+            snapshot,
+            start_version,
+            end_version,
+            compaction_path: compaction_path.location,
+        })
+    }
+
+    /// Get the path where the compaction file will be written
+    pub fn compaction_path(&self) -> &Url {
+        &self.compaction_path
+    }
+
+    /// Get an iterator over the compaction data to be written
+    ///
+    /// Performs action reconciliation for the version range specified in the constructor.
+    /// It creates a LogSegment that only includes commits in the specified version range,
+    /// and performs action reconciliation.
+    pub fn compaction_data(
+        &mut self,
+        engine: &dyn Engine,
+    ) -> DeltaResult<LogCompactionDataIterator> {
+        // Validate that the requested version range is within the snapshot's range
+        let snapshot_end_version = self.snapshot.version();
+        if self.end_version > snapshot_end_version {
+            return Err(Error::generic(format!(
+                "End version {} exceeds snapshot version {}",
+                self.end_version, snapshot_end_version
+            )));
+        }
+
+        // Create a log segment specifically for the compaction range
+        // This ensures we only process commits in [start_version, end_version]
+        let compaction_log_segment = LogSegment::for_table_changes(
+            engine.storage_handler().as_ref(),
+            self.snapshot.log_segment().log_root.clone(),
+            self.start_version,
+            Some(self.end_version),
+        )?;
+
+        // Read actions from the version-filtered log segment
+        let actions_iter = compaction_log_segment.read_actions(
+            engine,
+            COMPACTION_ACTIONS_SCHEMA.clone(),
+            COMPACTION_ACTIONS_SCHEMA.clone(),
+            None, // No predicate - we want all actions in the version range
+        )?;
+
+        let min_file_retention_timestamp_millis = self.deleted_file_retention_timestamp()?;
+
+        // Create checkpoint log replay processor for compaction
+        // This reuses the same reconciliation logic as checkpoints
+        let processor = CheckpointLogReplayProcessor::new(
```

Comment:
nice that we get to reuse this! should we add a few docs to `CheckpointLogReplayProcessor` that it's actually checkpoint + log compaction now? or is there another name we could come up with? maybe a mega-nit - can just take this as an improvement to consider later